src/storage.rs

```diff
@@ -419,200 +450,324 @@ impl Cache {
         reg.index.crate_(name)
     }
 
-    pub fn fetch_packages<'a>(
-        &mut self,
-        network: Option<Arc<Network>>,
-        packages: &[(PackageStr<'a>, &'a Version)],
-    ) -> Result<SortedMap<PackageStr<'a>, SortedMap<&'a Version, PathBuf>>, VetError> {
-        let _span = trace_span!("fetch-packages").entered();
-        // Don't do anything if we're mocked, or there is no work to do
-        if self.root.is_none() || packages.is_empty() {
-            return Ok(SortedMap::new());
-        }
-
-        let root = self.root.as_ref().unwrap();
-        let package_src_dir = root.join(TEMP_REGISTRY_SRC);
-        let package_cache_dir = root.join(TEMP_REGISTRY_CACHE);
-        let cargo_registry = self.cargo_registry.as_ref();
-
-        let mut paths = SortedMap::<PackageStr, SortedMap<&Version, PathBuf>>::new();
-        let mut to_download = Vec::new();
-
-        // Get all the cached things / find out what needs to be downloaded.
-        for (name, version) in packages {
-            let path = if **version == resolver::ROOT_VERSION {
-                // Empty package
-                root.join(TEMP_EMPTY_PACKAGE)
-            } else {
-                // First try to get a cached copy from cargo's register or our own
-                let dir_name = format!("{}-{}", name, version);
-
-                let fetched_src = cargo_registry
-                    .map(|reg| reg.src().join(&dir_name))
-                    .filter(|path| fetch_is_ok(path))
-                    .unwrap_or_else(|| package_src_dir.join(&dir_name));
-
-                if !fetch_is_ok(&fetched_src) {
-                    // If we don't have a cached copy, push this to the download queue
-                    let crate_name = format!("{}.crate", dir_name);
-                    let fetched_package = package_cache_dir.join(crate_name);
-                    to_download.push((name, version, fetched_package, fetched_src.clone()));
-                }
-
-                // Either this path exists or we'll download it, either way, it's right
-                fetched_src
-            };
+    #[tracing::instrument(skip(self, network))]
+    pub async fn fetch_package(
```

Comment:
there's a strange irony in async making it *better* for concurrency to define things as individual ops instead of bulk ops. I was trying to be responsible! Ah well, at least it made it easy to port the callers to joins.